tools/rpkg/R/backend-dbplyr__duckdb_connection.R

```diff
@@ -0,0 +1,383 @@
+#' DuckDB SQL backend for dbplyr
+#'
+#' @description
+#' This is a SQL backend for dbplyr tailored to take into account DuckDB's
+#' possibilities. This mainly follows the backend for PostgreSQL, but
+#' contains more mapped functions.
+#'
+#' @name backend-duckdb
+#' @aliases NULL
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' con <- DBI::dbConnect(duckdb::duckdb(), path = ":memory:")
+#'
+#' x <- tibble(txt = c("why", "video", "cross", "extra", "deal", "authority"))
+#' dbx <- copy_to(con, x, overwrite = TRUE)
+#'
+#' # x %>% filter(grepl("^.[hrx]", txt))
+#' dbx %>% filter(grepl("^.[hrx]", txt))
+#'
+#' # x %>% mutate(a = stringr::str_pad(txt, 10, side = "both", pad = ">"))
+#' dbx %>% mutate(a = str_pad(txt, 10, side = "both", pad = ">"))
+#'
+#' # x %>% mutate(a = stringr::str_replace_all(txt, "[aeiou]", "?"))
+#' dbx %>% mutate(a = str_replace_all(txt, "[aeiou]", "?"))
+#'
+#' DBI::dbDisconnect(con, shutdown = TRUE)
+#' duckdb::duckdb_shutdown(duckdb::duckdb())
+NULL
+
+#' Connection object for simulation of the SQL generation without actual database.
+#' dbplyr overrides database specific identifier and string quotes
+#' @param ... Any parameters to be forwarded
+#' @export
+#' @rdname backend-duckdb
+simulate_duckdb <- function(...) {
+  structure(list(), ..., class = c("duckdb_connection", "TestConnection", "DBIConnection"))
+}
+
+#' Connection object for simulation of the SQL generation without actual database.
+#' This version keeps the database specific identifier and string quotes, i.e.
+#' allows to translate to DuckDB SQL dialect.
+#' @param ... Any parameters to be forwarded
+#' @export
+#' @rdname backend-duckdb
+translate_duckdb <- function(...) {
+  structure(list(), ..., class = c("duckdb_connection", "DBIConnection"))
+}
+
+#' Declare which version of dbplyr API is being called.
+#' @param con A \code{\link{dbConnect}} object, as returned by \code{dbConnect()}
+#' @export
+dbplyr_edition.duckdb_connection <- function(con) {
+  2L
+}
+
+#' Description of the database connection
+#' @param con A \code{\link{dbConnect}} object, as returned by \code{dbConnect()}
+#' @name db_connection_describe
+#' @return
+#' String consisting of DuckDB version, user login name, operating system, R version and the name of database
+#' @export
+db_connection_describe.duckdb_connection <- function(con) {
+  info <- DBI::dbGetInfo(con)
+  paste0(
+    "DuckDB ", info$db.version, " [", Sys.info()["login"], "@",
+    paste(Sys.info()[c("sysname", "release")], collapse = " "), ":",
+    "R ", R.version$major, ".", R.version$minor, "/", info$dbname, "]"
+  )
+}
+
+duckdb_grepl <- function(pattern, x, ignore.case = FALSE, perl = FALSE, fixed = FALSE, useBytes = FALSE) {
+  # https://duckdb.org/docs/sql/functions/patternmatching
+  if (any(c(perl, fixed, useBytes))) {
+    stop("Parameters `perl`, `fixed` and `useBytes` in grepl are not currently supported in DuckDB backend", .call = FALSE)
+  }
```

Comment:
DuckDB seems to support [POSIX-style regular expressions](https://duckdb.org/docs/sql/functions/patternmatching). So the current default appears to be fine and is consistent with R.

The checks for arguments are [directly from PostgreSQL-backend in dbplyr](https://github.com/tidyverse/dbplyr/blob/912bb234e051321d6a4f361277b2a22980428147/R/backend-postgres.R#L46-L51). I would not change those until there is an agreement that we want "better" checks than in backends included in dbplyr. In addition, checking of arguments should then be as comprehensive for all translated functions, but that would require a lot of work and does not seem to be the practice in dbplyr backends at the moment. Maybe those could be added later on and especially focusing on the ones that users would like to have more info or better translation?!

It may also be possible to support `fixed` comparisons, and those could be added later on, if there appears to be need for such in real use cases.

Reply:
Found that DuckDB uses `RE2` as a regexp engine (#1310).